File: make_sample2barcode/make_sample2barcode.pl
#!/bin/env perl

=head1 NAME

make_sample2barcode.pl - create sample2barcode file from Excel spreadsheet

=head1 SYNOPSIS

B<make_sample2barcode> I<Excel sample key>

=head1 REQUIRES

Perl 5, 
File::Basename,
Tk, 
Tk::DropSite, 
Spreadsheet::Read, 
Spreadsheet::ParseExcel,
Spreadsheet::XLSX

=cut

use strict;
use warnings;

use Tk;
use Tk::DropSite;
use Cwd qw(abs_path);
use File::Basename;
use File::Spec;
use Spreadsheet::Read;
use Carp;
use Data::Dumper;

=head1 DESCRIPTION

Convert a Excel sample key to a STAMP sample2barcode.txt file.

=head2 Input file

The input file must be an Excel file with the name of the stamp run
in the format STAMP# (case-insensitive) somewhere in the file and
a table with columns: Name, lab# or acc#, mrn# and barcode (case-insensitive).
The control has no lab# or mrn#.

=head2 Output

The output is a tab-delimited sample2barcode_STAMP###.txt file for use 
with the STAMP analysis pipeline.  The output file will be written in
the same folder as the input file.

=head1 OPTIONS

=over 4

=item B<-h>, B<--help>

Print short usage message.

=back

=cut

### CONSTANTS

my $VERSION = "1.2";
my $BUILD = "160317";

#----------------

local $SIG{__WARN__} = sub {
    my $message = shift;
    croak "Error: " . $message;
 };

my $DefaultText = "";
my $TextField = "";
my $UserInput;
my $MainWindow;
my @Frames;
my $TextBox;
my $InputType;
my %Widgets = ();


sub setup_file_input {
    my $frame = shift;
    my $textfieldref = shift;
    my $l = $frame->Label(-width=>15, -text=>"STAMP sample key: ");
    my $e = $frame->Entry(-width=>50, -textvariable=>$textfieldref);
    $e->DropSite(-dropcommand => [\&accept_drop, $e ],
                -droptypes => ($^O eq 'MSWin32' ? 'Win32' : ['XDND', 'Sun']));
    my $b = $frame->Button(-width=>8, -text=>'Select file', 
                    -command=>[\&select_file_callback, $textfieldref]);
    $frame->pack();
    $l->pack(-side=>'left');
    $e->pack(-side=>'left');
    $b->pack(-side=>'left');
    return;
}

my $OUTDIR = undef;
if (@ARGV) {
    ($UserInput)  = @ARGV;
    if ($UserInput =~ /^--?h(elp)*$/i) {
        usage_message();
    } elsif ($UserInput =~ /^--?o[utdir]*$/i) {
        shift @ARGV;
        $OUTDIR = shift @ARGV;
        run_program(@ARGV);
    } else {
        print "Running program\n";
        run_program(@ARGV);
    }
} elsif ($#ARGV < 0){#run as GUI
    $MainWindow = MainWindow->new(-title=>'STAMP sample2barcode generator'.
                                          " (v$VERSION)");
    $MainWindow->geometry("620x470+0+0");

    $MainWindow->Label(-text=> "")->pack();
    my $textframe = $MainWindow->Frame->pack(-ipadx=>10, -ipady=>10);
    my $fileframe = $textframe->Frame->pack(-side=>'top');
    setup_file_input($fileframe, \$TextField);

    my $frame2 = $MainWindow->Frame;
    $frame2->Button(-width=>15, -text=>'Run', 
                    -command=>\&run_in_GUI)->pack(-side=>'left');
    $frame2->pack(-ipadx=>5, -ipady=>10);

    $TextBox = $MainWindow->Scrolled(qw' Text -height 20 -width 70', 
                                     -scrollbars=>'e')->pack;
    $TextBox->DropSite(-dropcommand => [\&accept_drop, $TextBox ],
              -droptypes => ($^O eq 'MSWin32' ? 'Win32' : ['XDND', 'Sun']));

    my $frame3 = $MainWindow->Frame;
    $frame3->Button( -width=>15, -text=>'Reset',
                     -command=>sub { 
                           undef $UserInput;
                           $TextField = $DefaultText;
                           $TextBox->delete("1.0",'end'); })
        ->pack(-side=>'left', -expand=>1);
    $frame3->Button(-width=>15, -text=>'Exit', 
                    -command=>sub { exit })->pack(-side=>'right', -expand=>1);
    $frame3->pack(-ipadx=>5, -ipady=>5);
    MainLoop();
}
else { usage_message() }


sub usage_message {
    print "Usage: $0 <STAMP sample key>\n";
    print "version $VERSION\n";
    exit;
}

sub run_program {
    my @inputs = @_;

    my $statuslist;
    my $fail = "FAIL";
    my $pass = "SUCCESS";

    if (@inputs and -f $inputs[0]) {
        foreach my $input (@inputs) {
            eval {
                my $message = create_message_string(abs_path($input));
                $statuslist->{$pass}->{$input} = $message;
            };
            if ($@){
                print "FAIL $@.\n";
                $statuslist->{$fail}->{$input} = $@;
            }
        }
    } else {
        print "No file to process.\n";
        $statuslist->{$fail}->{$inputs[0]} = "No file to process.";
    }
    return $statuslist;
}

sub create_message_string {
    my $inputfile = shift;
    my $test = shift || 0;

    my ($runnum, $data) = get_sample_data($inputfile);
    my $contents = create_sample2barcode($runnum, $data);
    # If script is run from the K drive, then save data to
    # same directory as script; otherwise, save to same directory
    # as input file.
    my ($basename, $outpath) = fileparse($inputfile);
    if ($0 =~ /^K/) { ($basename, $outpath) = fileparse($0); }
    if ($OUTDIR && -d $OUTDIR) { $outpath = $OUTDIR }
    my $outfile = File::Spec->catfile($outpath, 
                  "sample2barcode_STAMP$runnum.txt");
    print STDERR "\nOutput file: $outfile\n\n";
    open(my $ofh, ">", $outfile) or die ">$outfile: $!";
    print $ofh $contents;
    close $ofh;
    return $contents . "\nContents written to $outfile\n";
}

sub get_sample_data {
    my $inputfile = shift;

    print "Reading $inputfile\n";
    unless ($inputfile =~ /\.xls/) {
        die "ERROR: Input $inputfile not an Excel file\n";
    }
    my $wkbook = ReadData($inputfile) or
        die "Failed to read $inputfile: $!\n";
    my @cells = grep @$_, @{ $$wkbook[1]{cell} }; 
    my %columns = (
            NAME => -1,
            LAB => -1,
            MRN => -1,
            BARCODE => -1);
    unless (@cells) {
        die "No data in wkbook\n";
    }
    my $runnum = '';
    my %data;
    for(my $i=0; $i<@cells; $i++) {
        my @columnvals = map { defined $_ ? $_ : '' } @{ $cells[$i] };
        # Find stamp run
        my $stamprun_patt = '^stamp\s*[id:\s]*[\s_]*(\d+)([a-z]*)[\s_]*\b.*$';
        if (my @runnum = grep(/$stamprun_patt/i, @columnvals)) {
            if (!$runnum) { # take first matching candidate
                $runnum[0] =~ s/$stamprun_patt/$1/i;
                $runnum = sprintf "%03d", $runnum[0];
                if ($2 && length($2)<5) { $runnum .= $2 }
            }

        }
        if (grep(/name/i, @columnvals)) {
            $data{name} = get_column_values('name', \@columnvals);
        } elsif (grep(!/old/ && /(lab#|acc#)/i, @columnvals)) {
            $data{lab} = get_column_values('(lab#|acc#)', \@columnvals);
        } elsif (grep(/mrn#/i, @columnvals)) {
            $data{mrn} = get_column_values('mrn#', \@columnvals);
        } elsif (grep(/^\s*barcode\s*$/i, @columnvals)) {
            $data{barcode} = get_column_values('barcode', \@columnvals);
        }
    }
    unless (exists $data{name}) { die "Name column not found\n"; }
    unless (exists $data{lab}) { die "Lab column not found\n"; }
    unless (exists $data{mrn}) { die "MRN column not found\n"; }
    unless (exists $data{barcode}) { die "Barcode column not found\n"; }
#    print Dumper(\@cells);
#    print Dumper([\%data]);
    return ($runnum, \%data);
}

sub get_column_values {
    my $field = shift;
    my $columnvals = shift;

    my $flag = 0;
    my %values;
    # Get values in column after field
    my $match;
    for(my $j=0; $j<@$columnvals; $j++) {
        if ($flag && $$columnvals[$j]) {
            $values{$j} = $$columnvals[$j];
            $values{$j} =~ s/^\s+|\s+$//;
        } elsif ($$columnvals[$j] =~ /($field)/i) {
            $match = $1;
            $flag = 1;
        }
    }
    print STDERR "  $match: ".scalar(keys %values). " values\n";
    (\%values);
}

sub create_sample2barcode {
    my $runnum = shift;
    my $data = shift;

    my $contents = '';
    my @i = sort {$a<=>$b} keys %{ $$data{name} };

    my %samplenames;
    foreach my $i (@i) {
        my $name = $$data{name}{$i};
        next unless $name;
        my $lab = $$data{lab}{$i} || '';
        my $mrn = $$data{mrn}{$i} || '';
        my $sample;
        if ($name =~ /^(tr?u?q.?3|molt4|hd753).*$runnum/i) { 
            $sample = $name;
        } elsif ($name =~ /^tr?u?q.?3/i) {
            $sample = "TruQ3_$runnum";
        } elsif ($name =~ /^molt4/i) {
            $sample = "MOLT4_$runnum";
        } elsif ($name =~ /^hd753/i) {
            $sample = "HD753_$runnum";
        } elsif ($lab || $mrn) { #patient name
            # Change name to last name + first initial(s)
            $name =~ s/-//g; #remove hyphens in hyphenated names
            $name =~ s/,/_/g;
            my ($last, $first) = $name =~ /_/ ? split(/_/, $name, 2) :
                                 split(/\s/, $name, 2);
            unless (defined($first)) { $first = ''; }
            $first =~ s/([A-Z])[a-z]*/$1/g;
            $name = $last.$first;
            $name =~ s/[\s,]+//g;
            $name =~ s/_([A-Z])[^a-z]?/$1/;
            $name =~ s/\(/_/g;
            $name =~ s/\)//g;
            print STDERR "$$data{name}{$i} --> $name\n";
            $sample = join("_", $name, $lab, $mrn);
        } else { # no lab or mrn
            $sample = $name;
        }
        #remove spaces and non-standard characters
        $sample =~ s/[\s\x9F-\xFF]//g;
        #make sure not to have duplicate names
        if (exists $samplenames{$sample}) {
            $samplenames{$sample}++;
            $sample .="-$samplenames{$sample}";
        } else {
            $samplenames{$sample} = 1;
        }
        my $barcode = $$data{barcode}{$i};
        unless ($barcode) {
            print STDERR "No barcode for '$name'; SKIPPING\n";
            next;
        }
        $contents .= "$sample\t$barcode\n";
    }
    print STDERR "\n".$contents;
    ($contents);
}

#-----------------------------------------------------------------------------

sub run_in_GUI {
    unless ($UserInput) {
        $TextBox->delete("1.0",'end');
        $TextBox->insert("1.0", "Please enter a valid expression report.");
        return;
    }
    my $statuslist;
    eval {
        my @inputs = ($UserInput); 
        $statuslist = run_program(@inputs);
    };
    if ($@) {
        $TextBox->delete("1.0",'end');
        $TextBox->insert("end", "Status: failed!\n");
        $TextBox->insert("end", "\n");
        $TextBox->insert("end", "$@");
    }
    else {
        $TextBox->delete("1.0",'end');
        if ($statuslist) { ##print FAIL message first
            for my $status (sort {$a cmp $b} keys %$statuslist){
                $TextBox->insert("end" , 
                                 "\n===============$status===============\n");
                for my $file (sort keys %{$statuslist->{$status}}){
                    $TextBox->insert("end", "Input file: $file\n\n");
                    if ($statuslist->{$status}->{$file}) {
                        $TextBox->insert("end", "\n".
                                    $statuslist->{$status}->{$file}."\n\n\n");
                    }
                }
                $TextBox->insert("end" , "\n");
            }
        }
    }
}


sub select_file_callback  {
    my $selected = $MainWindow->getOpenFile(
        -defaultextension => ".xlsx",
        -filetypes => [
                [ 'Excel files', ['.xlsx','.xls']],
                ['All Files', '*']
            ],
        -title => "Open File"
    );
    if ($selected) {
        $UserInput = $selected;
        $TextField = $selected;
        $TextBox->delete("1.0",'end');
    }
}


sub accept_drop {
    my $widget = shift;
    my $selection = shift;

    my $filename;
    eval {
        if ($^O eq 'MSWin32') {
            $filename = $widget->SelectionGet(-selection => $selection,
                                              'STRING');
        } else {
            $filename = $widget->SelectionGet(-selection => $selection,
                                              'FILE_NAME');
        }
    };
    if (defined $filename) {
        $UserInput = $filename;
        $TextField = $filename;
        $TextBox->delete("1.0",'end');
    }
}

File: make_sample2barcode/make_sample2barcode.py
#!/usr/bin/env python

"""
make_sample2barcode.py

Create sample2barcode.txt file from Excel coversheet.
"""

import os
import sys
import openpyxl
import re
import traceback
import wx
import wx.richtext 
from collections import defaultdict
from argparse import ArgumentParser

PROGRAM=os.path.basename(sys.argv[0])
VERSION="2.2"
BUILD="180209"

## REVISION HISTORY
# v2.2 180209 - Fix display numbering when multiple inputs entered at once
# v2.1 170828 - Restrict sample names to letters, digits, hyphens, underscores
# v2.0 170518 - Convert from perl to python
# 170111 - Remove periods from sample names
# 170217 - Always remove commas from sample name


#----classes------------------------------------------------------------------

class STAMPCoversheet():
    def __init__(self, coversheet, outdir=None, debug=False):
        self.coversheet = coversheet
        self.runnum = None
        self.fields = None
        self.data = []
        self.outdir = None
        self.outfile = None
        self.debug = debug

        try:
            self._parse_coversheet()
            self.get_output_filename(outdir)
        except Exception as e:
            traceback.print_exc()

    def _parse_coversheet(self):
        stamprun_patt = re.compile(
            r'stamp\s*[id:\s]*[\s_]*(\d+)([a-z]*)[\s_]*\b', flags=re.I)
        wb = openpyxl.load_workbook(self.coversheet, read_only=True)
        wx = wb.active
        for row in wx.iter_rows():
            cells = [ cell.value for cell in row ]
            while cells and cells[-1]==None:
                cells.pop() # remove trailing blanks
            if self.fields and cells:
                # data lines have both Name and barcode fields
                d = dict(zip(self.fields, cells))
                if d.get('Name') and d.get('barcode'):
                    self.data.append(d)
            elif 'Name' in cells and 'lab#' in cells and 'mrn#' in cells:
                # column names
                self.fields = cells
            else: 
                # possible header line with run# info
                line = "\t".join([str(c) for c in cells]) + "\n"
                m = stamprun_patt.search(line)
                if m:
                    self.runnum = "{:03d}".format(int(m.group(1)))
                    if m.group(2) and len(m.group(2))<5:
                        self.runnum += m.group(2)

    def format_sample2barcode(self):
        self.sample2barcode = []
        samples_seen = {}
        for d in self.data:
            name = d['Name'].replace(',','_').replace('(','_').replace(')','')
            lab = str(d['lab#']).strip() if d.get('lab#') else ''
            mrn = str(d['mrn#']).strip() if d.get('mrn#') else ''
            barcode = d['barcode'].strip()
            if not mrn.isdigit():
                if mrn: 
                    print "WARNING: MRN not digit '{}'".format(mrn)
                mrn = ''
            # Check if control sample
            ctrl_patt = '(tr?u?q.?3|molt4|hd753).*{}'.format(self.runnum)
            if re.match(ctrl_patt, name, flags=re.I):
                sample = name
            # add run num to control name if not already present
            elif re.match('tr?u?q.?3', name, flags=re.I):
                sample = "TruQ3_{}".format(self.runnum)
            elif re.match('molt4', name, flags=re.I):
                sample = "MOLT4_{}".format(self.runnum)
            elif re.match('hd753', name, flags=re.I):
                sample = "HD753_{}".format(self.runnum)
            # Patient samples should have lab# or mrn#
            elif lab or mrn: 
                # Change name to last name + first initial(s)
                #  * Entry usually Last, First or Last_First
                #  * Research samples usually sample_research, so don't
                #    split into last, first if '_research'
#                names = re.split('[_,]+\s*(?!research)', name.replace('-', ''), 1)
                names = re.split('[_,]+\s*(?!research)', name, 1)
                last = names.pop(0)
                first = names.pop() if names else ''
                first = re.sub('([A-Z])[a-z]*', r'\1', first)
                name = re.sub('[\s,]+', '', last+first)
                if lab and lab in name: 
                    # if lab entry already in name, don't duplicate
                    lab = ''
                sample = '_'.join([ v.encode('utf-8') for v in (name, lab, mrn)])
            else:
                sample = name
            # remove spaces and non-standard characters
            # only allow English letters, digits, hyphens and underscores
            #sample = re.sub(r'[\s\.\x9F-\xFF]', '', sample)
            sample = re.sub(r'[^-\w]', '', sample)
            # make sure not to have duplicate names
            if sample in samples_seen:
                samples_seen[sample] += 1
                sample += '-{}'.format(samples_seen[sample])
            samples_seen[sample] = 1
            self.sample2barcode.append("{}\t{}".format(sample, barcode))
            if self.debug:
                print "{:50s}\t{:30s}\t{}".format(sample, name, d['Name'])
                sys.stdout.flush()
        return self.sample2barcode

    def get_output_filename(self, outdir=None):
        if not outdir:
            outdir = os.path.dirname(self.coversheet)
        if self.runnum:
            outfile = "sample2barcode_STAMP{}.txt".format(self.runnum)
        else:
            outfile = "sample2barcode.txt"
        self.outdir = outdir
        self.outfile = os.path.join(outdir, outfile)
        return self.outfile

    def write_sample2barcode_file(self, outfile=None):
        if not outfile:
            outfile = self.outfile
        if not self.sample2barcode:
            self.format_sample2barcode()
        with open(outfile, 'w') as ofh:
            ofh.write("\n".join(self.sample2barcode)+"\n")


#----gui.py-------------------------------------------------------------------

class Sample2Barcode_App(wx.App):
    def __init__(self, args, **kwargs):
        self.args = args
        wx.App.__init__(self, kwargs)

    def OnInit(self):
        self.frame = StampFrame(self.args)
        self.frame.Show()
        self.SetTopWindow(self.frame)
        return True

class StampRTC(wx.richtext.RichTextCtrl):
    def __init__(self, parent):
        wx.richtext.RichTextCtrl.__init__(self, parent, -1, "",
                        style=wx.TE_READONLY|wx.TE_MULTILINE|wx.HSCROLL)
        self.Bind(wx.EVT_MOUSE_EVENTS, self.DoNothing)

    def DoNothing(self, event):
        pass

    def AddIntroBlurb(self):
        intro_blurb="This script creates sample2barcode.txt files for"+\
                    " use with the STAMP analysis software."+\
                    " Input files must follow the STAMP coversheet conventions."
        intro_items = [ ]
        self.BeginFontSize(10)
        self.Newline()
        self.Newline()
        self.WriteText(intro_blurb)
        self.Newline()
        self.BeginSymbolBullet('*', 25, 30)
        for [label, descr] in intro_items:
            self.BeginBold()
            self.WriteText(label)
            self.EndBold()
            self.WriteText(' -- '+descr)
            self.Newline()
        self.EndSymbolBullet()
        self.EndFontSize()
        self.Newline()
        self.Newline()

class StampFrame(wx.Frame):
    def __init__(self, args):
        self.args = args
        wx.Frame.__init__(self, None, size=(550,500),
                          title="{} v{}".format(PROGRAM,VERSION))

        panel = wx.Panel(self)
        label = wx.StaticText(panel, -1, "Drop STAMP coversheets here:")
#        self.rtc = wx.richtext.RichTextCtrl(panel,-1, "",
#                        style=wx.TE_READONLY|wx.TE_MULTILINE|wx.HSCROLL)
        self.rtc = StampRTC(panel)
        self.rtc.AddIntroBlurb()
        button_quit = wx.Button(panel, -1, "Quit", style=wx.BU_EXACTFIT)
        button_quit.SetToolTip(wx.ToolTip("Quit application"))
        self.Bind(wx.EVT_BUTTON, self.OnCloseMe, button_quit)
        self.Bind(wx.EVT_CLOSE, self.OnCloseWindow)

        sizer = wx.BoxSizer(wx.VERTICAL)
        sizer.Add(label, 0, wx.ALL, 5)
        sizer.Add(self.rtc, 1, wx.EXPAND|wx.ALL, 5)

        button_sizer = wx.BoxSizer(wx.HORIZONTAL)
        button_sizer.AddStretchSpacer()
        button_sizer.Add(button_quit, 0, wx.ALIGN_CENTER_VERTICAL)
        sizer.Add(button_sizer, 0, wx.ALL|wx.EXPAND, 5)
        panel.SetSizer(sizer)

        filedrop = FileDropProcessing(self.rtc, self.args)
        self.rtc.SetDropTarget(filedrop)

    def OnCloseMe(self, event):
        self.Close(True)

    def OnCloseWindow(self, event):
        self.Destroy()
        
class FileDropProcessing(wx.FileDropTarget):
    def __init__(self, window, args):
        wx.FileDropTarget.__init__(self)
        self.window = window
        self.args = args
        self.num_samples = 0
        self.current_pos = 0

    def ScrollWindow(self):
        pos = self.window.GetScrollRange(wx.VERTICAL)
        self.window.Scroll(0, pos)

    def WriteFormattedText(self, boldtext='', normaltext='', bullet=False,
                           newline=True):
        self.window.MoveEnd()
        if self.current_pos:
            self.window.SetCaretPosition(self.current_pos)
        if bullet:
            self.window.BeginSymbolBullet('*', 25, 30)
        if boldtext:
            self.window.BeginBold()
            self.window.WriteText(boldtext)
            self.window.EndBold()
        if normaltext:
            self.window.WriteText(normaltext)
        if newline: 
            self.window.Newline()
        if bullet:
            self.window.EndSymbolBullet()
        self.ScrollWindow()
        self.current_pos = self.window.GetCaretPosition()

    def OnDropFiles(self, x, y, coversheets):
        coversheet_data = []
        num = 0
        for coversheet in coversheets:
            s2bdata = STAMPCoversheet(coversheet, debug=self.args.debug)
            if s2bdata.fields:
                coversheet_data.append(s2bdata)
            else:
                self.WriteFormattedText(normaltext="Not a recognized input file:  {}".\
                    format(os.path.basename(coversheet)))
                sys.stderr.write("  WARNING: Unrecognized format {}\n".format(coversheet))
            num += 1
            label = " {}".format(num) if len(coversheets)>1 else ''
            self.WriteFormattedText("\nCoversheet{}: ".format(label), 
                                    os.path.basename(coversheet))
            sys.stdout.write("\nCoversheet {}\n".format(coversheet))
            formatted_data = s2bdata.format_sample2barcode()
            if not formatted_data:
                self.WriteFormattedText(newline=False,
                    normaltext="No data in {}\n".format(os.path.basename(coversheet)))
                sys.stdout.write("  WARNING: No data {}\n".format(s2bdata.coversheet))
            else:
                self.WriteFormattedText(normaltext="    "+"\n    ".join(formatted_data))
                self.WriteFormattedText(newline=False,
                    normaltext="Writing {}\n".format(s2bdata.outfile))
                sys.stdout.write("  Writing {}\n".format(s2bdata.outfile))
                s2bdata.write_sample2barcode_file()

def run_gui(args):
    app = Sample2Barcode_App(args)
    app.MainLoop()

#-----------------------------------------------------------------------------
if __name__=='__main__':
    descr = "Create sample2barcode.txt file from STAMP cover sheet."
    parser = ArgumentParser(description=descr)
    parser.add_argument("coversheets", nargs="*",
                        help="STAMP excel coversheets")
    parser.add_argument("-o", "--outdir", 
                        help="Directory to save output file(s)")
    parser.add_argument("--debug", default=False, action='store_true',
                        help="Write debugging messages")

    args = parser.parse_args()
    if not args.coversheets:
        run_gui(args)
    else:
        for coversheet in args.coversheets:
            sys.stdout.write("\nCoversheet {}\n".format(coversheet))
            s2bdata = STAMPCoversheet(coversheet, outdir=args.outdir,
                                      debug=args.debug)
            if not s2bdata.fields:
                sys.stderr.write("  WARNING: Unrecognized format {}\n".\
                format(coversheet))
            else:
                formatted_data = s2bdata.format_sample2barcode()
                if not formatted_data:
                    sys.stdout.write("  WARNING: No data {}\n".format(coversheet))
                else:
                    sys.stdout.write("  Writing {}\n".format(s2bdata.outfile))
                    s2bdata.write_sample2barcode_file()


            



